File: Data Structures Algorithns/Old/LinkedList.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node
{
    int data;
    struct Node *next;
} Node;

int getChoiceFromMenu();
int performAction(Node **, int);
void insertBeginning(Node **, int);
void insertEnd(Node **, int);
void insert(Node **, int, int);
void print(Node *);

// Bitch Lasagna
int main(int argc, char const *argv[])
{
    printf("LINKED LIST");
    Node *head = NULL;
    int runMore = 1;
    while (runMore)
        runMore = performAction(&head, getChoiceFromMenu());
    return 0;
}

int getChoiceFromMenu()
{
    int choice = 69420;
    printf("\n1. Insert Element at Beginning\n");
    printf("2. Insert Element at End\n");
    printf("3. Insert Element at Position\n");
    printf("4. Display List\n");
    printf("5. Exit\n > ");
    scanf("%d", &choice);
    return choice;
}

int performAction(Node **head, int choice)
{
    int runMore = 1;
    int n, i;
    switch (choice)
    {
    case 1:
        printf("\nEnter element to insert: > ");
        scanf("%d", &n);
        insertBeginning(head, n);
        break;
    case 2:
        printf("\nEnter element to insert: > ");
        scanf("%d", &n);
        insertEnd(head, n);
        break;
    case 3:
        printf("\nEnter element to insert: > ");
        scanf("%d", &n);
        printf("\nEnter position to insert: > ");
        scanf("%d", &i);
        insert(head, n, i);
    case 4:
        print(*head);
        break;
    case 5:
        runMore = 0;
        break;
    default:
        break;
    }
    return runMore;
}

void insertBeginning(Node **head, int data)
{
    Node *temp = (Node *)malloc(sizeof(Node));
    temp->data = data;
    temp->next = *head != NULL ? *head : NULL;
    *head = temp;
}
void insertEnd(Node **head, int data)
{
    Node *temp = (Node *)malloc(sizeof(Node));
    temp->data = data;
    temp->next = NULL;
    Node *current = *head != NULL ? *head : NULL;
    while (current->next != NULL)
    {
        current = current->next;
    }

    current->next = temp;
}
void insert(Node **head, int data, int iter)
{
    Node *temp = (Node *)malloc(sizeof(Node));
    temp->data = data;
    temp->next = NULL;
    Node *current = *head;
    for (int i = 0; i < iter; i++)
    {
        current = current->next;
    }

    current->next = temp;
}

void print(Node *head)
{
    if (head == NULL)
    {
        printf("List is Empty\n");
        return;
    }
    else
    {
        do
        {
            printf("%d ", head->data);
            head = head->next;
        } while (head->next != NULL);
    }
}

File: Data Structures Algorithns/Old/linkedList.py
class Node:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next


class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

    def append(self, data):
        node = Node(data)
        if self.head:
            current = self.head
            while current.next:
                current = current.next
            current.next = node
        else:
            self.head = node
        self.size += 1

    def display(self,seperator=',', newLine=True):
        current = self.head
        while current:
            print(current.data,end=seperator)
            current = current.next
        if newLine: print()

    def delete(self, data):
        if self.head.data == data:
            self.head = self.head.next
            self.size -= 1
        else:
            current = self.head
            while current.next:
                if current.next.data == data:
                    current.next = current.next.next
                    self.size -= 1
                    break
                current = current.next

    def insert(self, data, index):
        node = Node(data)
        current = self.head
        for _ in range(index):
            current = current.next
        node.next = current.next
        current.next = node
    
    def search(self, data):
        current = self.head
        while current:
            if current.data == data: return True
            current = current.next
        return False
    def sort(self):
        while True:
            current = self.head
            while current.next:
                if current.data > current.next.data:
                    current.data, current.next.data = current.next.data, current.data
                    break
                current = current.next
            else:
                break

linkedList = LinkedList()
linkedList.append(3)
linkedList.append(1)
linkedList.append(4)
linkedList.append(5)
linkedList.append(2)
linkedList.display()

linkedList.delete(3)
linkedList.insert(12,2)
linkedList.delete(1)
linkedList.display()

print("Sorting:")
linkedList.sort()
linkedList.display()

print(linkedList.search(int(input("Enter Element to Search For: > "))))
